File: packages/react-native/ios/CrumbReactNative.swift
import CrumbCore
import CrumbUI
import Foundation
import NitroModules

final class CrumbReactNative: HybridCrumbReactNativeSpec {
    private let decoder = JSONDecoder()
    private let logBuffer = ReactNativeLogBuffer()

    func start(configurationJson: String) throws {
        let data = Data(configurationJson.utf8)
        let payload = try decoder.decode(ConfigurationPayload.self, from: data)
        let bundle = Bundle.main
        let appVersion = payload.release?.appVersion
            ?? bundle.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String
            ?? "0"
        let nativeBuild = payload.release?.nativeBuild
            ?? bundle.object(forInfoDictionaryKey: "CFBundleVersion") as? String
            ?? "0"
        let logOptions = payload.diagnostics?.logs ?? .init()

        logBuffer.configure(
            lookback: logOptions.lookbackMs.map { TimeInterval(milliseconds: $0) } ?? 60,
            maximumEntries: logOptions.maximumEntries ?? 200,
            maximumBytes: logOptions.maximumBytes ?? 65_536
        )

        try Crumb.start(
            CrumbConfiguration(
                projectKey: payload.projectKey,
                environment: payload.environment,
                release: CrumbRelease(
                    appVersion: appVersion,
                    nativeBuild: nativeBuild,
                    bundleVersion: payload.release?.bundleVersion
                ),
                invocation: Set(
                    (payload.invocation ?? [.shake, .programmatic]).map(\.native)
                ),
                capture: CrumbCaptureOptions(
                    screenshot: payload.capture?.screenshot ?? true,
                    maximumScreenshotDimension:
                        payload.capture?.maximumScreenshotDimension ?? 2_048,
                    maximumScreenshotBytes:
                        payload.capture?.maximumScreenshotBytes ?? 5_242_880
                ),
                diagnostics: CrumbDiagnosticsOptions(
                    healthCheckURL: try payload.diagnostics?.healthCheckURL(),
                    timeout: payload.diagnostics?.timeoutMs
                        .map { TimeInterval(milliseconds: $0) } ?? 2,
                    logs: CrumbLogOptions(
                        enabled: logOptions.enabled ?? true,
                        lookback: logOptions.lookbackMs
                            .map { TimeInterval(milliseconds: $0) } ?? 60,
                        maximumEntries: logOptions.maximumEntries ?? 200,
                        maximumBytes: logOptions.maximumBytes ?? 65_536,
                        provider: logBuffer
                    )
                ),
                privacy: CrumbPrivacyOptions(
                    maskAllTextInputs: payload.privacy?.maskAllTextInputs ?? true,
                    maskScreenshotsBeforeUpload:
                        payload.privacy?.maskScreenshotsBeforeUpload ?? true
                ),
                reporter: CrumbReporterOptions(
                    theme: payload.reporter?.theme?.native ?? .system,
                    visibleFields: Set(
                        (payload.reporter?.visibleFields ?? [.category, .description]).map(\.native)
                    )
                ),
                evidence: Set(
                    (payload.evidence ?? EvidencePayload.allCases).map(\.native)
                ),
                application: CrumbApplicationMetadata(name: payload.application?.name),
                customContext: CrumbCustomContextOptions(
                    values: payload.customContext?.values ?? [:],
                    allowedKeys: Set(payload.customContext?.allowedKeys ?? [])
                ),
                workspacePolicy: CrumbWorkspacePolicyOptions(
                    url: try payload.workspacePolicy?.url(),
                    timeout: payload.workspacePolicy?.timeoutMs
                        .map { TimeInterval(milliseconds: $0) } ?? 2
                ),
                upload: CrumbUploadOptions(
                    ingestionURL: try payload.upload?.ingestionURL()
                )
            )
        )
    }

    func installReporter() throws -> Promise<Bool> {
        Promise.async { @MainActor in
            Crumb.installReporter()
        }
    }

    func show() throws -> Promise<Bool> {
        Promise.async { @MainActor in
            Crumb.show()
        }
    }

    func addLog(entryJson: String) throws {
        let data = Data(entryJson.utf8)
        let entry = try decoder.decode(LogEntryPayload.self, from: data)
        logBuffer.append(entry.native)
    }

    func clearLogs() throws {
        logBuffer.clear()
    }
}

private struct ConfigurationPayload: Decodable {
    let projectKey: String
    let environment: String
    let release: ReleasePayload?
    let invocation: [InvocationPayload]?
    let capture: CapturePayload?
    let diagnostics: DiagnosticsPayload?
    let privacy: PrivacyPayload?
    let reporter: ReporterPayload?
    let evidence: [EvidencePayload]?
    let application: ApplicationPayload?
    let customContext: CustomContextPayload?
    let workspacePolicy: WorkspacePolicyPayload?
    let upload: UploadPayload?
}

private struct ReleasePayload: Decodable {
    let appVersion: String?
    let nativeBuild: String?
    let bundleVersion: String?
}

private enum InvocationPayload: String, Decodable {
    case shake
    case programmatic

    var native: CrumbInvocation {
        switch self {
        case .shake: .shake
        case .programmatic: .programmatic
        }
    }
}

private struct CapturePayload: Decodable {
    let screenshot: Bool?
    let maximumScreenshotDimension: Int?
    let maximumScreenshotBytes: Int?
}

private struct DiagnosticsPayload: Decodable {
    let healthCheckUrl: String?
    let timeoutMs: Int?
    let logs: LogOptionsPayload?

    func healthCheckURL() throws -> URL? {
        guard let healthCheckUrl else { return nil }
        guard let url = URL(string: healthCheckUrl) else {
            throw BridgeConfigurationError.invalidURL("healthCheckUrl")
        }
        return url
    }
}

private struct LogOptionsPayload: Decodable {
    let enabled: Bool?
    let lookbackMs: Int?
    let maximumEntries: Int?
    let maximumBytes: Int?

    init(
        enabled: Bool? = nil,
        lookbackMs: Int? = nil,
        maximumEntries: Int? = nil,
        maximumBytes: Int? = nil
    ) {
        self.enabled = enabled
        self.lookbackMs = lookbackMs
        self.maximumEntries = maximumEntries
        self.maximumBytes = maximumBytes
    }
}

private struct PrivacyPayload: Decodable {
    let maskAllTextInputs: Bool?
    let maskScreenshotsBeforeUpload: Bool?
}

private struct ReporterPayload: Decodable {
    let theme: ThemePayload?
    let visibleFields: [ReporterFieldPayload]?
}

private enum ThemePayload: String, Decodable {
    case system
    case light
    case dark

    var native: CrumbTheme {
        switch self {
        case .system: .system
        case .light: .light
        case .dark: .dark
        }
    }
}

private enum ReporterFieldPayload: String, Decodable {
    case category
    case description

    var native: CrumbReporterField {
        switch self {
        case .category: .category
        case .description: .description
        }
    }
}

private enum EvidencePayload: String, CaseIterable, Decodable {
    case screenshot
    case performance
    case network
    case logs
    case threadStacks = "thread_stacks"
    case healthCheck = "health_check"
    case customContext = "custom_context"

    var native: CrumbEvidenceCategory {
        switch self {
        case .screenshot: .screenshot
        case .performance: .performance
        case .network: .network
        case .logs: .logs
        case .threadStacks: .threadStacks
        case .healthCheck: .healthCheck
        case .customContext: .customContext
        }
    }
}

private struct ApplicationPayload: Decodable {
    let name: String?
}

private struct CustomContextPayload: Decodable {
    let values: [String: String]?
    let allowedKeys: [String]?
}

private struct WorkspacePolicyPayload: Decodable {
    let urlString: String?
    let timeoutMs: Int?

    enum CodingKeys: String, CodingKey {
        case urlString = "url"
        case timeoutMs
    }

    func url() throws -> URL? {
        guard let urlString else { return nil }
        guard let url = URL(string: urlString) else {
            throw BridgeConfigurationError.invalidURL("workspacePolicy.url")
        }
        return url
    }
}

private struct UploadPayload: Decodable {
    let ingestionUrl: String?

    func ingestionURL() throws -> URL? {
        guard let ingestionUrl else { return nil }
        guard let url = URL(string: ingestionUrl) else {
            throw BridgeConfigurationError.invalidURL("ingestionUrl")
        }
        return url
    }
}

private enum BridgeConfigurationError: Error {
    case invalidURL(String)
}

private struct LogEntryPayload: Decodable {
    let timestampMs: Int64
    let level: String
    let source: String
    let category: String
    let message: String

    var native: CrumbLogEntry {
        CrumbLogEntry(
            timestamp: Date(timeIntervalSince1970: TimeInterval(timestampMs) / 1_000),
            level: level.nativeLogLevel,
            source: source,
            category: category,
            message: message
        )
    }
}

private extension String {
    var nativeLogLevel: CrumbLogLevel {
        switch self {
        case "debug": .debug
        case "notice": .notice
        case "warning": .warning
        case "error": .error
        case "fault": .fault
        default: .info
        }
    }
}

private final class ReactNativeLogBuffer: CrumbLogProvider, @unchecked Sendable {
    private let lock = NSLock()
    private var entries: [CrumbLogEntry] = []
    private var lookback: TimeInterval = 60
    private var maximumEntries = 200
    private var maximumBytes = 65_536

    func configure(
        lookback: TimeInterval,
        maximumEntries: Int,
        maximumBytes: Int
    ) {
        withLock {
            self.lookback = lookback
            self.maximumEntries = maximumEntries
            self.maximumBytes = maximumBytes
            prune(now: Date())
        }
    }

    func append(_ entry: CrumbLogEntry) {
        withLock {
            entries.append(entry)
            prune(now: Date())
        }
    }

    func clear() {
        withLock {
            entries.removeAll(keepingCapacity: false)
        }
    }

    func recentLogs() throws -> [CrumbLogEntry] {
        withLock {
            prune(now: Date())
            return entries
        }
    }

    private func prune(now: Date) {
        let earliest = now.addingTimeInterval(-lookback)
        entries.removeAll { $0.timestamp < earliest }
        while entries.count > maximumEntries || byteCount(entries) > maximumBytes {
            entries.removeFirst()
        }
    }

    private func byteCount(_ entries: [CrumbLogEntry]) -> Int {
        entries.reduce(into: 0) { total, entry in
            total += entry.source.utf8.count
                + entry.category.utf8.count
                + entry.message.utf8.count
                + 32
        }
    }

    private func withLock<Result>(_ action: () throws -> Result) rethrows -> Result {
        lock.lock()
        defer { lock.unlock() }
        return try action()
    }
}

private extension TimeInterval {
    init(milliseconds: Int) {
        self = TimeInterval(milliseconds) / 1_000
    }
}

File: packages/react-native/android/src/main/java/com/margelo/nitro/crumbsdk/reactnative/CrumbReactNative.kt
package com.margelo.nitro.crumbsdk.reactnative

import android.app.Application
import android.os.Build
import com.facebook.react.bridge.UiThreadUtil
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import dev.crumb.core.Crumb
import dev.crumb.core.CrumbCaptureOptions
import dev.crumb.core.CrumbConfiguration
import dev.crumb.core.CrumbDiagnosticsOptions
import dev.crumb.core.CrumbEvidenceCategory
import dev.crumb.core.CrumbInvocation
import dev.crumb.core.CrumbLogEntry
import dev.crumb.core.CrumbLogLevel
import dev.crumb.core.CrumbLogOptions
import dev.crumb.core.CrumbLogProvider
import dev.crumb.core.CrumbReporterField
import dev.crumb.core.CrumbReporterOptions
import dev.crumb.core.CrumbPrivacyOptions
import dev.crumb.core.CrumbRelease
import dev.crumb.core.CrumbTheme
import dev.crumb.core.CrumbUploadOptions
import dev.crumb.core.CrumbApplicationMetadata
import dev.crumb.core.CrumbCustomContextOptions
import dev.crumb.core.CrumbWorkspacePolicyOptions
import dev.crumb.ui.CrumbReporter
import org.json.JSONObject

class CrumbReactNative : HybridCrumbReactNativeSpec() {
    private val logBuffer = ReactNativeLogBuffer()
    @Volatile
    private var javascriptCrashCaptureEnabled = false

    override fun start(configurationJson: String) {
        val context = requireNotNull(NitroModules.applicationContext) {
            "React Native is not ready. Call Crumb.start after the app has mounted."
        }
        val payload = JSONObject(configurationJson)
        val release = payload.optJSONObject("release") ?: JSONObject()
        val diagnostics = payload.optJSONObject("diagnostics")
        val logs = diagnostics?.optJSONObject("logs")
        val packageInfo = context.packageManager.getPackageInfo(context.packageName, 0)
        val nativeBuild = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            packageInfo.longVersionCode.toString()
        } else {
            @Suppress("DEPRECATION")
            packageInfo.versionCode.toString()
        }
        val maximumEntries = logs?.optionalInt("maximumEntries") ?: 200
        val maximumBytes = logs?.optionalInt("maximumBytes") ?: 65_536
        val lookbackMillis = logs?.optionalLong("lookbackMs") ?: 60_000L

        logBuffer.configure(
            lookbackMillis = lookbackMillis,
            maximumEntries = maximumEntries,
            maximumBytes = maximumBytes,
        )

        Crumb.start(
            CrumbConfiguration(
                projectKey = payload.getString("projectKey"),
                environment = payload.getString("environment"),
                release = CrumbRelease(
                    appVersion = release.optionalString("appVersion")
                        ?: packageInfo.versionName
                        ?: "0",
                    nativeBuild = release.optionalString("nativeBuild") ?: nativeBuild,
                    bundleVersion = release.optionalString("bundleVersion"),
                ),
                invocation = payload.invocations(),
                capture = payload.captureOptions(),
                diagnostics = CrumbDiagnosticsOptions(
                    healthCheckUrl = diagnostics?.optionalString("healthCheckUrl"),
                    timeoutMillis = diagnostics?.optionalLong("timeoutMs") ?: 2_000,
                    logs = CrumbLogOptions(
                        enabled = logs?.optionalBoolean("enabled") ?: true,
                        lookbackMillis = lookbackMillis,
                        maximumEntries = maximumEntries,
                        maximumBytes = maximumBytes,
                        provider = logBuffer,
                    ),
                ),
                privacy = payload.privacyOptions(),
                upload = CrumbUploadOptions(
                    ingestionUrl = payload.optJSONObject("upload")
                        ?.optionalString("ingestionUrl"),
                ),
                reporter = payload.reporterOptions(),
                evidence = payload.evidenceOptions(),
                application = payload.applicationMetadata(),
                customContext = payload.customContextOptions(),
                workspacePolicy = payload.workspacePolicyOptions(),
            ),
        )
        javascriptCrashCaptureEnabled = payload.optJSONObject("javascriptCrashCapture")
            ?.optionalBoolean("enabled") == true
    }

    override fun canCollectLogs(): Boolean = Crumb.canCollectLogs()

    override fun installReporter(): Promise<Boolean> {
        val promise = Promise<Boolean>()
        val context = runCatching {
            requireNotNull(NitroModules.applicationContext).applicationContext
        }.getOrElse {
            promise.reject(it)
            return promise
        }
        val shouldRecoverJavaScriptCrashes = javascriptCrashCaptureEnabled
        Thread({
            if (shouldRecoverJavaScriptCrashes) {
                runCatching { Crumb.recoverPendingJavaScriptCrashes(context) }
            }
            UiThreadUtil.runOnUiThread {
                runCatching {
                    CrumbReporter.install(context as Application)
                }.onSuccess(promise::resolve)
                    .onFailure(promise::reject)
            }
        }, "Crumb JavaScript crash recovery").start()
        return promise
    }

    override fun show(): Promise<Boolean> {
        val promise = Promise<Boolean>()
        UiThreadUtil.runOnUiThread {
            runCatching {
                val context = requireNotNull(NitroModules.applicationContext)
                val activity = context.getCurrentActivity() ?: return@runCatching false
                CrumbReporter.show(activity)
            }.onSuccess(promise::resolve)
                .onFailure(promise::reject)
        }
        return promise
    }

    override fun recordJavaScriptCrash(payloadJson: String) {
        if (!javascriptCrashCaptureEnabled) return
        val context = requireNotNull(NitroModules.applicationContext)
        val payload = JSONObject(payloadJson)
        Crumb.recordJavaScriptCrash(context, payload.javascriptCrash())
    }

    override fun addLog(entryJson: String) {
        if (!Crumb.canCollectLogs()) return
        val payload = JSONObject(entryJson)
        logBuffer.append(
            CrumbLogEntry(
                timestampMillis = payload.getLong("timestampMs"),
                level = payload.getString("level").toNativeLogLevel(),
                source = payload.optString("source", "react-native"),
                category = payload.optString("category", "javascript"),
                message = payload.getString("message"),
            ),
        )
    }

    override fun clearLogs() {
        logBuffer.clear()
    }
}

private fun JSONObject.invocations(): Set<CrumbInvocation> {
    val values = optJSONArray("invocation")
        ?: return setOf(CrumbInvocation.SHAKE, CrumbInvocation.PROGRAMMATIC)
    return buildSet {
        for (index in 0 until values.length()) {
            when (values.getString(index)) {
                "shake" -> add(CrumbInvocation.SHAKE)
                "programmatic" -> add(CrumbInvocation.PROGRAMMATIC)
            }
        }
    }
}

private fun JSONObject.captureOptions(): CrumbCaptureOptions {
    val capture = optJSONObject("capture") ?: return CrumbCaptureOptions()
    return CrumbCaptureOptions(
        screenshot = capture.optionalBoolean("screenshot") ?: true,
        maximumScreenshotDimension = capture.optionalInt("maximumScreenshotDimension") ?: 2_048,
        maximumScreenshotBytes = capture.optionalInt("maximumScreenshotBytes") ?: 5_242_880,
    )
}

private fun JSONObject.privacyOptions(): CrumbPrivacyOptions {
    val privacy = optJSONObject("privacy") ?: return CrumbPrivacyOptions()
    return CrumbPrivacyOptions(
        maskAllTextInputs = privacy.optionalBoolean("maskAllTextInputs") ?: true,
        maskScreenshotsBeforeUpload = privacy.optionalBoolean("maskScreenshotsBeforeUpload") ?: true,
    )
}

private fun JSONObject.reporterOptions(): CrumbReporterOptions {
    val reporter = optJSONObject("reporter") ?: return CrumbReporterOptions()
    val theme = when (reporter.optionalString("theme") ?: "system") {
        "system" -> CrumbTheme.SYSTEM
        "light" -> CrumbTheme.LIGHT
        "dark" -> CrumbTheme.DARK
        else -> error("reporter.theme is unsupported")
    }
    val fields = reporter.optJSONArray("visibleFields")?.let { array ->
        buildSet {
            for (index in 0 until array.length()) {
                when (array.getString(index)) {
                    "category" -> add(CrumbReporterField.CATEGORY)
                    "description" -> add(CrumbReporterField.DESCRIPTION)
                    else -> error("reporter.visibleFields contains an unsupported value")
                }
            }
        }
    } ?: CrumbReporterOptions().visibleFields
    return CrumbReporterOptions(theme = theme, visibleFields = fields)
}

private fun JSONObject.evidenceOptions(): Set<CrumbEvidenceCategory> {
    val evidence = optJSONArray("evidence") ?: return CrumbEvidenceCategory.entries.toSet()
    return buildSet {
        for (index in 0 until evidence.length()) {
            when (evidence.getString(index)) {
                "screenshot" -> add(CrumbEvidenceCategory.SCREENSHOT)
                "performance" -> add(CrumbEvidenceCategory.PERFORMANCE)
                "network" -> add(CrumbEvidenceCategory.NETWORK)
                "logs" -> add(CrumbEvidenceCategory.LOGS)
                "thread_stacks" -> add(CrumbEvidenceCategory.THREAD_STACKS)
                "health_check" -> add(CrumbEvidenceCategory.HEALTH_CHECK)
                "custom_context" -> add(CrumbEvidenceCategory.CUSTOM_CONTEXT)
                else -> error("evidence contains an unsupported value")
            }
        }
    }
}

private fun JSONObject.applicationMetadata(): CrumbApplicationMetadata {
    val application = optJSONObject("application") ?: return CrumbApplicationMetadata()
    return CrumbApplicationMetadata(name = application.optionalString("name"))
}

private fun JSONObject.customContextOptions(): CrumbCustomContextOptions {
    val customContext = optJSONObject("customContext") ?: return CrumbCustomContextOptions()
    val values = customContext.optJSONObject("values")?.let { valuesObject ->
        buildMap {
            val names = valuesObject.keys()
            while (names.hasNext()) {
                val name = names.next()
                put(name, valuesObject.getString(name))
            }
        }
    } ?: emptyMap()
    val allowedKeys = customContext.optJSONArray("allowedKeys")?.let { array ->
        buildSet {
            for (index in 0 until array.length()) add(array.getString(index))
        }
    } ?: emptySet()
    return CrumbCustomContextOptions(values = values, allowedKeys = allowedKeys)
}

private fun JSONObject.workspacePolicyOptions(): CrumbWorkspacePolicyOptions {
    val policy = optJSONObject("workspacePolicy") ?: return CrumbWorkspacePolicyOptions()
    return CrumbWorkspacePolicyOptions(
        url = policy.optionalString("url"),
        timeoutMillis = policy.optionalLong("timeoutMs") ?: 2_000,
    )
}

private fun JSONObject.javascriptCrash(): dev.crumb.core.CrumbJavaScriptCrash {
    val kind = when (getString("kind")) {
        "fatal_exception" -> dev.crumb.core.CrumbJavaScriptCrashKind.FATAL_EXCEPTION
        "unhandled_rejection" -> dev.crumb.core.CrumbJavaScriptCrashKind.UNHANDLED_REJECTION
        else -> error("javascript crash kind is unsupported")
    }
    val source = when (optionalString("source") ?: "javascript") {
        "javascript" -> dev.crumb.core.CrumbJavaScriptCrashSource.JAVASCRIPT
        "native_termination_wrapper" ->
            dev.crumb.core.CrumbJavaScriptCrashSource.NATIVE_TERMINATION_WRAPPER
        else -> error("javascript crash source is unsupported")
    }
    val breadcrumbs = optJSONArray("breadcrumbs")?.let { array ->
        buildList {
            for (index in 0 until array.length()) {
                val breadcrumb = array.optJSONObject(index) ?: continue
                add(
                    dev.crumb.core.CrumbJavaScriptBreadcrumb(
                        timestampMillis = breadcrumb.optLong("timestampMs", System.currentTimeMillis()),
                        category = breadcrumb.optString("category", "javascript"),
                        message = breadcrumb.optString("message", ""),
                    ),
                )
            }
        }
    } ?: emptyList()
    return dev.crumb.core.CrumbJavaScriptCrash(
        kind = kind,
        source = source,
        errorType = getString("errorType"),
        message = getString("message"),
        rawStack = optionalString("rawStack"),
        fingerprint = optionalString("fingerprint"),
        occurredAtMillis = optLong("occurredAtMs", System.currentTimeMillis()),
        nativeTerminationWrapper = optBoolean("nativeTerminationWrapper", false),
        breadcrumbs = breadcrumbs,
    )
}

private fun JSONObject.optionalString(name: String): String? =
    if (has(name) && !isNull(name)) getString(name) else null

private fun JSONObject.optionalBoolean(name: String): Boolean? =
    if (has(name) && !isNull(name)) getBoolean(name) else null

private fun JSONObject.optionalInt(name: String): Int? =
    if (has(name) && !isNull(name)) getInt(name) else null

private fun JSONObject.optionalLong(name: String): Long? =
    if (has(name) && !isNull(name)) getLong(name) else null

private fun String.toNativeLogLevel(): CrumbLogLevel = when (this) {
    "debug" -> CrumbLogLevel.DEBUG
    "info" -> CrumbLogLevel.INFO
    "notice" -> CrumbLogLevel.NOTICE
    "warning" -> CrumbLogLevel.WARNING
    "error" -> CrumbLogLevel.ERROR
    "fault" -> CrumbLogLevel.FAULT
    else -> CrumbLogLevel.INFO
}

private class ReactNativeLogBuffer : CrumbLogProvider {
    private val lock = Any()
    private val entries = ArrayDeque<CrumbLogEntry>()
    private var lookbackMillis = 60_000L
    private var maximumEntries = 200
    private var maximumBytes = 65_536

    fun configure(
        lookbackMillis: Long,
        maximumEntries: Int,
        maximumBytes: Int,
    ) = synchronized(lock) {
        this.lookbackMillis = lookbackMillis
        this.maximumEntries = maximumEntries
        this.maximumBytes = maximumBytes
        prune(System.currentTimeMillis())
    }

    fun append(entry: CrumbLogEntry) = synchronized(lock) {
        entries.addLast(entry)
        prune(System.currentTimeMillis())
    }

    fun clear() = synchronized(lock) {
        entries.clear()
    }

    override fun recentLogs(): List<CrumbLogEntry> = synchronized(lock) {
        prune(System.currentTimeMillis())
        entries.toList()
    }

    private fun prune(now: Long) {
        val earliest = now - lookbackMillis
        while (entries.firstOrNull()?.timestampMillis?.let { it < earliest } == true) {
            entries.removeFirst()
        }
        while (entries.size > maximumEntries || byteCount() > maximumBytes) {
            entries.removeFirstOrNull()
        }
    }

    private fun byteCount(): Int = entries.sumOf { entry ->
        entry.source.toByteArray().size +
            entry.category.toByteArray().size +
            entry.message.toByteArray().size +
            32
    }
}
